### PR TITLE
Add protocol components and data flow documentation

### PR DESCRIPTION
**Pull request by jonastheis**
[rendered](https://github.com/iotaledger/goshimmer/blob/feat/protocol-components-doc/docs/protocol_specification/components.md)

### CODE REVIEW

**Review by capossele — APPROVED** [review]

**Review by jkrvivian — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -32,30 +32,30 @@
...
 ## Client-Library and HTTP API reference
 
 You can use the Go client-library to interact with GoShimmer (located under `github.com/iotaledger/goshimmer/client`).
 
-You can find more info about this on our [client-lib](https://goshimmer.docs.iota.org/apis/api.html) and [Web API](https://goshimmer.docs.iota.org/apis/webAPI.html) GitHub page.
+You can find more info about this on our [client-lib](https://goshimmer.docs.iota.org/apis/api.html) and [Web API](https://goshimmer.docs.iota.org/apis/webAPI.html) documentation.
 
 ## Getting started
 
 You can find tutorials on how to [setup a GoShimmer node](https://goshimmer.docs.iota.org/tutorials/setup.html), [writing a dApp](https://goshimmer.docs.iota.org/tutorials/dApp.html), [obtaining tokens from the faucet](https://goshimmer.docs.iota.org/tutorials/request_funds.html) and more on our [GitHub Page](https://goshimmer.docs.iota.org/goshimmer.html).
```
Comment (apenzk) [inline README.md:54]:
```suggestion
You can find tutorials on how to [setup a GoShimmer node](https://goshimmer.docs.iota.org/tutorials/setup.html), [writing a dApp](https://goshimmer.docs.iota.org/tutorials/dApp.html), [obtaining tokens from the faucet](https://goshimmer.docs.iota.org/tutorials/request_funds.html) and more in the [GoShimmer documentation](https://goshimmer.docs.iota.org/goshimmer.html).
```

File: README.md [hunk 2]
```diff
@@ -32,30 +32,30 @@
...
 
 ### Compiling from source
 
-We always recommend to run your node via [Docker](https://goshimmer.docs.iota.org/tutorials/setup.html). However, you can also compile the source and run the node from the compiled binary. GoShimmer uses [RocksDB](https://github.com/linxGnu/grocksdb) as its underlying db engine. That requires a few dependencies before building the project: 
+We always recommend running your node via [Docker](https://goshimmer.docs.iota.org/tutorials/setup.html). However, you can also compile the source and run the node from the compiled binary. GoShimmer uses [RocksDB](https://github.com/linxGnu/grocksdb) as its underlying db engine. That requires a few dependencies before building the project: 
 - librocksdb
 - libsnappy
 - libz
```
Comment (apenzk) [inline README.md:61]: We should change line 65 to 
Please follow this [guide](https://github.com/facebook/rocksdb/blob/master/INSTALL.md) to build the above libs.

File: docs/protocol_specification/components.md [hunk 3]
```diff
@@ -1,44 +1,97 @@
...
 
-### Node identities and Mana
+![](layers.png)
 
-Nodes receive rewards for being good actors in a system that secures the network and provides Sybil protection. 
+## Network Layer
+The network is maintained by the network layer modules, which can be characterized as a pure P2P overlay network, meaning that it is a system that runs on top of another network (e.g., the internet), and where all nodes have the same roles and perform the same actions (in contrast to client-server systems). GoShimmer's Network Layer consists of three basic modules: the [peer discovery](./autopeering.md#peer-discovery) module (which provides a list of nodes actively using the network), and the [neighbor selection](./autopeering.md#neighbor-selection) module (also known as autopeering), which actually select peers. Finally, the P2P Communication manages a node's neighbors, either selected via [autopeering](autopeering.md) or [manual peering](../tutorials/manual_peering.md).
```
Comment (apenzk) [inline docs/protocol_specification/components.md:7]:
```suggestion
The network is maintained by the network layer modules, which can be characterized as a pure P2P overlay network, meaning that it is a system that runs on top of another network (e.g., the internet), and where all nodes have the same roles and perform the same actions (in contrast to client-server systems). GoShimmer's Network Layer consists of three basic modules: the [peer discovery](./autopeering.md#peer-discovery) module (which provides a list of nodes actively using the network), and the [neighbor selection](./autopeering.md#neighbor-selection) module (also known as autopeering), which actually selects peers. Finally, the P2P Communication manages a node's neighbors, either selected via [autopeering](autopeering.md) or [manual peering](../tutorials/manual_peering.md).
```

File: docs/protocol_specification/components.md [hunk 4]
```diff
@@ -1,44 +1,97 @@
...
 
-For more details see the [GoShimmer mana specs](./mana.md).
+## Communication Layer
+The communication layer concerns the information propagated through the network layer, which is contained in objects called messages. This layer forms a DAG with messages as vertices called [The Tangle](./tangle.md): a replicated, shared and distributed data structure that emerges through a combination of deterministic rules, cooperation, and (either direct or virtual) voting—as FPC and approval weight based finality.
```
Comment (apenzk) [inline docs/protocol_specification/components.md:10]:
```suggestion
The communication layer concerns the information propagated through the network layer, which is contained in objects called messages. This layer forms a DAG with messages as vertices called the [Tangle](./tangle.md): a replicated, shared and distributed data structure that emerges—through a combination of deterministic rules, cooperation, and (either direct or virtual) voting—as FPC and approval weight based finality.
```

File: docs/protocol_specification/components.md [hunk 5]
```diff
@@ -1,44 +1,97 @@
...
+Additionally, all nodes must run what we call the consensus applications, which regulate timestamps in the messages and resolve conflicts.
+The consensus mechanism implemented in GoShimmer is leaderless and consists out of multiple components:
+1. The [Fast Probabilistic Consensus (FPC)](./consensus_mechanism.md#fpc) provides pre-consensus in form of a binary voting protocol that produces consensus on a bit and uses a [distributed Random Number Generator (dRNG)](./consensus_mechanism.md#drng) to be resilient against metastable states.
+2. A virtual [voting protocol (Approval Weight)](./consensus_mechanism.md#approval-weight-aw) that provides finality similarly to the longest chain rule in Nakamoto consensus (i.e., heaviest branch) for branches and messages.
 
-The concept of payloads creates infinite possibilities while also improving data processing efficiency. 
 
-### Autopeering
+## Data flow - Overview
```
Comment (apenzk) [inline docs/protocol_specification/components.md:22]:
```suggestion
## Data Flow - Overview
```

File: docs/protocol_specification/components.md [hunk 6]
```diff
@@ -1,44 +1,97 @@
...
-A mechanism for automatically finding neighbors to build a robust ever-changing network.
- 
-### Timestamps
+As an example, take the Parser component. The function `ProcessGossipMessage` will trigger the method `Parse`, which is the only entry to the component. There are three possible outcomes to the `Parser`: triggering a `ParsingFailed` event, a `MessageRejected` event, or a `MessageParsed` event. In the last case, the event will trigger the `StoreMessage` method (which is the entry to the Storage component), whereas the first two events do not trigger any other component.
 
-A declared time variable for messages that allows message ordering, rate control and identifies malicious behavior. 
+![](data-flow.png)
 
-### Tip selection
+We call this the data flow, i.e., the life cycle of a message, from message reception (meaning that we focus here on the point of view of a node receiving a message issued by another node) up until acceptance in the Tangle. Notice that any message, either created locally by the node or received from a neighbor needs to pass through the data flow.
```
Comment (apenzk) [inline docs/protocol_specification/components.md:29]:
```suggestion
We call this the data flow, i.e., the [life cycle of a message](../protocol_specification/protocol.html), from message reception (meaning that we focus here on the point of view of a node receiving a message issued by another node) up until acceptance in the Tangle. Notice that any message, either created locally by the node or received from a neighbor needs to pass through the data flow.
```

File: docs/protocol_specification/components.md [hunk 7]
```diff
@@ -1,44 +1,97 @@
...
 
-Faster, more secure and simpler. The new tip selection is not part of consensus anymore, but it allows us to reach finality within seconds.
+### Message Factory
+The IssuePayload function creates a valid payload which is provided to the CreateMessage method, along with a set of parents chosen with the Tip Selection Algorithm. Then, the Message Factory component is responsible to find a nonce compatible with the PoW requirements defined by the rate control module. Finally, the message is signed. Notice that the message generation should follow the rates imposed by the rate setter, as defined in [rate setting](./congestion_control.md#rate-setting).
```
Comment (apenzk) [inline docs/protocol_specification/components.md:32]:
```suggestion
The IssuePayload function creates a valid payload which is provided to the `CreateMessage` method, along with a set of parents chosen with the Tip Selection Algorithm. Then, the Message Factory component is responsible to find a nonce compatible with the PoW requirements defined by the rate control module. Finally, the message is signed. Notice that the message generation should follow the rates imposed by the rate setter, as defined in [rate setting](./congestion_control.md#rate-setting).
```

File: docs/protocol_specification/components.md [hunk 8]
```diff
@@ -1,44 +1,97 @@
...
+1. Recently Seen Bytes: it compares the incoming messages with a pool of recently seen bytes to filter duplicates.
+2. PoW check: it checks if the PoW requirements are met, currently set to the message hash starting with 22 zeroes.
 
-### Ledger state
+Followed by the bytes filters, the received bytes are parsed into a message and its corresponding payload and [syntactically validated](./tangle.md#syntactical-validation). From now on, the filters operate on message objects rather than just bytes.
 
-The Unspent Transaction Output (UTXO) model together with the Parallel Reality Ledger State enables nodes to efficiently keep track and update all the balances while managing potential conflicts.
+**Message filter**:
+1. Signature check: it checks if the message signature is valid.
+2. [Timestamp Difference Check for transactions](./tangle.md#message -timestamp-vs-transaction-timestamp): it checks if the timestamps of the payload, and the message are consistent with each other
```
Comment (apenzk) [inline docs/protocol_specification/components.md:45]:
```suggestion
2. [Timestamp Difference Check for transactions](./tangle.md#message-timestamp-vs-transaction-timestamp): it checks if the timestamps of the payload, and the message are consistent with each other
```

**Pull request merged**